src/azure-cli/azure/cli/command_modules/aro/tests/latest/custom_preparers.py

```diff
@@ -0,0 +1,111 @@
+    def __init__(
+        self,
+        name_prefix="clitest",
+        skip_assignment=True,
+        parameter_name="client_id",
+        parameter_password="client_secret",
+        dev_setting_sp_name="AZURE_CLI_TEST_DEV_SP_NAME",
+        dev_setting_sp_password="AZURE_CLI_TEST_DEV_SP_PASSWORD",
+        key="aro_csp",
+    ):
+        super(AROClusterServicePrincipalPreparer, self).__init__(
+            name_prefix, 
+            skip_assignment,
+            parameter_name,
+            parameter_password,
+            dev_setting_sp_name,
+            dev_setting_sp_password,
+            key,
+        )
+        self.aad = AADManager(self.cli_ctx)
+        self.client_id_to_replace = None
+        self.client_secret_to_replace = None
+        self.delete_app_on_completion = False
+
+    def create_resource(self, name, **kwargs):
+        if self.live_test or self.test_class_instance.in_recording:
+            if self.dev_setting_sp_name:
+                client_id = self.dev_setting_sp_name
+                client_secret = self.dev_setting_sp_password
+            else:
+                subscription = self.test_class_instance.get_subscription_id()
+                resource_group = self.test_class_instance.kwargs.get('rg')
+                command = 'az ad sp create-for-rbac -n {} --role contributor --scopes "{}"'\
+                    .format(name, resource_id(subscription=subscription, resource_group=resource_group))
+                
+                try:
+                    self.result = self.live_only_execute(self.cli_ctx, command).get_output_in_json()
+                except AttributeError:
+                    pass
+
+
+                client_id = self.result['appId']
+                client_secret = self.result.get('password') or GraphClientPasswordReplacer.PWD_REPLACEMENT
+
+                self.delete_app_on_completion = True
+
+            # call AbstractPreparer.moniker to make resource counts and self.resource_moniker consistent between live
+            # and play-back. see SingleValueReplacer.process_request, AbstractPreparer.__call__._preparer_wrapper
+            # and ScenarioTest.create_random_name. This is so that when self.create_random_name is called for the
+            # first time during live or playback, it would have the same value.
+            # In short, the default sp preparer in live mode does not call moniker, which leads to inconsistent counts.
+            _ = self.moniker
+
+            self.client_id_to_replace = client_id
+            self.client_secret_to_replace = client_secret
+
+        else:
+            client_id = MOCK_GUID
+            client_secret = MOCK_SECRET
+
+        self.test_class_instance.kwargs[self.key] = client_id
+        self.test_class_instance.kwargs["{}_pass".format(self.key)] = client_secret
+
+        return {
+            self.parameter_name: client_id,
+            self.parameter_password: client_secret,
+        }
```

Comment:
IMHO this function could be broken up a bit. I got lost with all the if and elses + indentation. If we break that in chunks it is a bit easier to read + understand and we can even delete some elses to reduce nesting.

I would propose something like that (in case you find it useful):
```python
    def create_resource(self, name, **kwargs):
        client_id, client_secret = self.configureAndGetClientInfo(name)

        self.test_class_instance.kwargs[self.key] = client_id
        self.test_class_instance.kwargs["{}_pass".format(self.key)] = client_secret

        return {
            self.parameter_name: client_id,
            self.parameter_password: client_secret,
        }

    def configureAndGetClientInfo(self, name):
        if self.live_test or self.test_class_instance.in_recording:
            client_id, client_secret = self.configureBasedOnDevSettingsAndGetClientInfo(name)

            # call AbstractPreparer.moniker to make resource counts and self.resource_moniker consistent between live
            # and play-back. see SingleValueReplacer.process_request, AbstractPreparer.__call__._preparer_wrapper
            # and ScenarioTest.create_random_name. This is so that when self.create_random_name is called for the
            # first time during live or playback, it would have the same value.
            # In short, the default sp preparer in live mode does not call moniker, which leads to inconsistent counts.
            _ = self.moniker

            self.client_id_to_replace = client_id
            self.client_secret_to_replace = client_secret
            return client_id,client_secret
            
        return MOCK_GUID, MOCK_SECRET 

    def configureBasedOnDevSettingsAndGetClientInfo(self, name):
        if self.dev_setting_sp_name:
            client_id = self.dev_setting_sp_name
            client_secret = self.dev_setting_sp_password
            
            return client_id, client_secret

        subscription = self.test_class_instance.get_subscription_id()
        resource_group = self.test_class_instance.kwargs.get('rg')
        command = 'az ad sp create-for-rbac -n {} --role contributor --scopes "{}"'\
                .format(name, resource_id(subscription=subscription, resource_group=resource_group)) 
        try:
            self.result = self.live_only_execute(self.cli_ctx, command).get_output_in_json()
        except AttributeError:
            pass

        client_id = self.result['appId']
        client_secret = self.result.get('password') or GraphClientPasswordReplacer.PWD_REPLACEMENT

        self.delete_app_on_completion = True
        
        return client_id,client_secret
```

If I had more context I would even break it a bit further but something like the above would be clear enough for me (I am sure the functions names I have proposed can be improved).

Reply:
This is a good suggestion, thank you! I'll be performing this refactor after I get some tests written for this class.